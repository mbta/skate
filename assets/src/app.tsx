// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
// tslint:disable-next-line
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "@babel/polyfill"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "leaflet/dist/leaflet.css"
import "phoenix_html"
import * as React from "react"
import ReactDOM from "react-dom"
import AppStateWrapper from "./components/appStateWrapper"

if (window.FS && window.username) {
  window.FS.identify(window.username, { displayName: window.username })
}

// Import local files
//
// Local files can be imported directly using relative paths, for example:
// import socket from "./socket"

ReactDOM.render(<AppStateWrapper />, document.getElementById("app"))
