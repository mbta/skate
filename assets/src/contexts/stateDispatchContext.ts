import { createContext } from "react"
import { Dispatch, initialState, State } from "../state"

type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export default StateDispatchContext
