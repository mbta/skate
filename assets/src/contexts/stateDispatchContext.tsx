import React, { createContext, ReactElement } from "react"
import { Dispatch, initialState, State } from "../state"

export type StateDispatchContextData = [State, Dispatch]

// tslint:disable-next-line: no-empty
const noopDispatch: Dispatch = () => {}
export const StateDispatchContext = createContext([
  initialState,
  noopDispatch,
] as StateDispatchContextData)

export const StateDispatchProvider = ({
  state,
  dispatch,
  children,
}: {
  state: State
  dispatch: Dispatch
  children: ReactElement<HTMLElement>
}) => {
  return (
    <StateDispatchContext.Provider value={[state, dispatch]}>
      {children}
    </StateDispatchContext.Provider>
  )
}
