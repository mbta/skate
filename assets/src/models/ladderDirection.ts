import { ByRouteId, DirectionId, RouteId, Timepoint } from "../schedule"

export enum LadderDirection {
  ZeroToOne,
  OneToZero,
}

const defaultLadderDirection: LadderDirection = LadderDirection.ZeroToOne

export type LadderDirections = ByRouteId<LadderDirection>

export const emptyLadderDirectionsByRouteId: LadderDirections = {}

export const getLadderDirectionForRoute = (
  ladderDirections: LadderDirections,
  routeId: RouteId
): LadderDirection => ladderDirections[routeId] || defaultLadderDirection

const flipLadderDirection = (
  ladderDirection: LadderDirection
): LadderDirection =>
  ladderDirection === LadderDirection.ZeroToOne
    ? LadderDirection.OneToZero
    : LadderDirection.ZeroToOne

export const flipLadderDirectionForRoute = (
  ladderDirections: LadderDirections,
  routeId: RouteId
): LadderDirections => {
  const currentLadderDirection = getLadderDirectionForRoute(
    ladderDirections,
    routeId
  )
  return {
    ...ladderDirections,
    [routeId]: flipLadderDirection(currentLadderDirection),
  }
}

export const orderTimepoints = (
  timepointsFromApi: Timepoint[],
  ladderDirection: LadderDirection
): Timepoint[] =>
  // Timepoints come from the API in the ZeroToOne direction
  // Use slice to make a copy of the array before destructively reversing
  ladderDirection === LadderDirection.OneToZero
    ? timepointsFromApi.slice().reverse()
    : timepointsFromApi

export enum VehicleDirection {
  Down,
  Up,
}

export const directionOnLadder = (
  directionId: DirectionId,
  ladderDirection: LadderDirection
): VehicleDirection =>
  (directionId === 1) === (ladderDirection === LadderDirection.ZeroToOne)
    ? VehicleDirection.Down
    : VehicleDirection.Up
