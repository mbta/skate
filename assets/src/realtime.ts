import {
  BlockId,
  DirectionId,
  RouteId,
  RoutePatternId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule"

import { Crowding } from "./models/crowding"
import { DetourId } from "./models/detoursList"

export interface BlockWaiver {
  startTime: Date
  endTime: Date
  causeId: number
  causeDescription: string
  remark: string | null
}

export interface DataDiscrepancy {
  attribute: string
  sources: DataDiscrepancySource[]
}

export interface DataDiscrepancySource {
  id: SourceId
  value: string | null
}

export type EndOfTripType = "another_trip" | "swing_off" | "pull_back"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  routePatternId: RoutePatternId | null
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  incomingTripDirectionId: DirectionId | null
  layoverDepartureTime: number | null
  scheduledTimepointStatus: VehicleTimepointStatus | null
  scheduledLogonTime: number | null
  routeStatus: RouteStatus
  blockWaivers: BlockWaiver[]
  currentPieceStartPlace: string | null
  currentPieceFirstRoute: RouteId | null
}

export type NotificationId = string

// A Notification's `content` is populated with an Elixir struct and the
// `__struct__` field is serialized so that the type can be discerned on the
// frontend.
export enum NotificationType {
  BridgeMovement = "Elixir.Notifications.Db.BridgeMovement",
  BlockWaiver = "Elixir.Notifications.Db.BlockWaiver",
  Detour = "Elixir.Notifications.Db.Detour",
}

export interface BridgeLoweredNotification {
  $type: NotificationType.BridgeMovement
  status: "lowered"
}

export interface BridgeRaisedNotification {
  $type: NotificationType.BridgeMovement
  status: "raised"
  loweringTime: Date
}

export type BridgeNotification =
  | BridgeLoweredNotification
  | BridgeRaisedNotification

export type BlockWaiverNotification = {
  $type: NotificationType.BlockWaiver
  createdAt: Date
  reason: BlockWaiverReason
  routeIds: RouteId[]
  runIds: RunId[]
  tripIds: TripId[]
  operatorId: string | null
  operatorName: string | null
  routeIdAtCreation: string | null
  startTime: Date
  endTime: Date | null
}

export enum DetourNotificationStatus {
  Activated = "activated",
  Deactivated = "deactivated",
}

export type DetourNotification = {
  $type: NotificationType.Detour
  status: DetourNotificationStatus
  detourId: DetourId
  headsign: string
  route: string
  direction: string
  origin: string
}

export type NotificationContentTypes =
  | BridgeNotification
  | BlockWaiverNotification
  | DetourNotification

export interface Notification<
  TNotification extends NotificationContentTypes = NotificationContentTypes
> {
  id: NotificationId
  createdAt: Date
  state: NotificationState
  content: TNotification
}

export function isBlockWaiverNotification(
  notification: Notification
): notification is Notification<BlockWaiverNotification> {
  return notification.content.$type === NotificationType.BlockWaiver
}
export type BlockWaiverReason =
  | "other"
  | "accident"
  | "adjusted"
  | "disabled"
  | "diverted"
  | "manpower"
  | "operator_error"
  | "traffic"

export type NotificationState = "unread" | "read" | "deleted"

export type SourceId = string

export type RouteStatus = "on_route" | "laying_over" | "pulling_out"

export type RunId = string

export interface ByRunId<T> {
  [runId: string]: T
}

export type TrainVehicleId = string

export interface TrainVehicle {
  id: TrainVehicleId
  latitude: number
  longitude: number
  bearing: number
}

export interface Vehicle {
  id: VehicleId
  label: string | null
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId | null
  routeId: RouteId | null
  routePatternId: RoutePatternId | null
  tripId: TripId | null
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string | null
  operatorFirstName: string | null
  operatorLastName: string | null
  operatorLogonTime: Date | null
  overloadOffset?: number
  bearing?: number
  blockId: BlockId | null
  previousVehicleId: string | null
  scheduleAdherenceSecs: number | null
  incomingTripDirectionId: DirectionId | null
  isShuttle: boolean
  isOverload: boolean
  isOffCourse: boolean
  isRevenue: boolean | null
  layoverDepartureTime: number | null
  pullbackPlaceName: string | null
  dataDiscrepancies: DataDiscrepancy[]
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  routeStatus: RouteStatus
  endOfTripType: EndOfTripType
  blockWaivers: BlockWaiver[]
  crowding: Crowding | null
}

export interface VehicleInScheduledService extends Vehicle {
  directionId: DirectionId
  routeId: RouteId
  blockId: BlockId
}

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId | null
  stopName: string | null
}

export interface VehicleScheduledLocation {
  routeId: RouteId
  routePatternId: RoutePatternId | null
  directionId: DirectionId
  tripId: TripId
  runId: RunId
  timeSinceTripStartTime: number
  headsign: string | null
  viaVariant: ViaVariant | null
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}
