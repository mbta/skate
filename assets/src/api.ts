import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
