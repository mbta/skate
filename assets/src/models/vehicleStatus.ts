import featureIsEnabled from "../laboratoryFeatures"
import { Vehicle } from "../realtime.d"

/** Where a vehicle is relative to its schedule.
 * Swiftly assigns this to all vehicles.
 */
export type OnTimeStatus = "on-time" | "early" | "late"

export enum HeadwaySpacing {
  VeryBunched = "very-bunched",
  Bunched = "bunched",
  Ok = "ok",
  Gapped = "gapped",
  VeryGapped = "very-gapped",
}

/** How the vehicle triangle should be drawn,
 * taking into account the vehicle's state and the settings
 */
export type DrawnStatus = OnTimeStatus | "off-course" | "ghost" | "plain"

export const onTimeStatus = (scheduleAdherenceSecs: number): OnTimeStatus => {
  const oneMinuteInSeconds = 60
  const sixMinutesInSeconds = 360

  if (scheduleAdherenceSecs < -oneMinuteInSeconds) {
    return "early"
  } else if (scheduleAdherenceSecs > sixMinutesInSeconds) {
    return "late"
  } else {
    return "on-time"
  }
}

export const drawnStatus = (vehicle: Vehicle): DrawnStatus => {
  if (vehicle.isOffCourse) {
    return "off-course"
  } else if (
    isShuttle(vehicle) ||
    (featureIsEnabled("headway_ladder_colors") &&
      vehicle.headwaySpacing !== null)
  ) {
    // Headway lines give the status instead of the vehicles
    return "plain"
  } else {
    return onTimeStatus(vehicle.scheduleAdherenceSecs)
  }
}

export const humanReadableScheduleAdherence = (vehicle: Vehicle): string =>
  vehicle.isOffCourse
    ? "Invalid"
    : humanReadableOnTimeStatus(onTimeStatus(vehicle.scheduleAdherenceSecs))

const humanReadableOnTimeStatus = (status: OnTimeStatus): string => {
  switch (status) {
    case "early":
      return "early"

    case "on-time":
      return "on time"

    case "late":
      return "late"
  }
}

export const humanReadableHeadwaySpacing = (
  spacing: HeadwaySpacing | null
): string => {
  switch (spacing) {
    case null:
      return "good"

    case HeadwaySpacing.VeryGapped:
      return "very gapped"

    case HeadwaySpacing.Gapped:
      return "gapped"

    case HeadwaySpacing.Ok:
      return "good"

    case HeadwaySpacing.Bunched:
      return "bunched"

    case HeadwaySpacing.VeryBunched:
      return "very bunched"
  }
}

export const headwaySpacingToString = (spacing: HeadwaySpacing): string => {
  switch (spacing) {
    case HeadwaySpacing.VeryGapped:
      return "very-gapped"

    case HeadwaySpacing.Gapped:
      return "gapped"

    case HeadwaySpacing.Ok:
      return "ok"

    case HeadwaySpacing.Bunched:
      return "bunched"

    case HeadwaySpacing.VeryBunched:
      return "very-bunched"
  }
}

export const statusClass = (status: DrawnStatus): string => {
  switch (status) {
    case "plain":
      return ""

    default:
      return status
  }
}

const isShuttle = (vehicle: Vehicle): boolean =>
  (vehicle.runId || "").startsWith("999")
