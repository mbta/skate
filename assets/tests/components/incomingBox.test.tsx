import { mount } from "enzyme"
import React from "react"
import renderer from "react-test-renderer"
import IncomingBox from "../../src/components/incomingBox"
import { StateDispatchProvider } from "../../src/contexts/stateDispatchContext"
import { LadderDirection } from "../../src/models/ladderDirection"
import { HeadwaySpacing } from "../../src/models/vehicleStatus"
import { Ghost, Vehicle } from "../../src/realtime"
import { initialState, selectVehicle } from "../../src/state"

import vehicleFactory from "../factories/vehicle"
import ghostFactory from "../factories/ghost"

describe("IncomingBox", () => {
  test("renders empty state", () => {
    const tree = renderer
      .create(
        <IncomingBox
          vehiclesAndGhosts={[]}
          ladderDirection={LadderDirection.ZeroToOne}
          selectedVehicleId={undefined}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a vehicle", () => {
    const vehicle: Vehicle = {
      id: "y0654",
      label: "0654",
      runId: "126-1056",
      timestamp: 1576091706,
      latitude: 42.36296,
      longitude: -71.05814,
      directionId: 0,
      routeId: "111",
      tripId: "42199995",
      headsign: "Woodlawn",
      viaVariant: "5",
      operatorId: "1",
      operatorFirstName: "OPPIE",
      operatorLastName: "OPERATOR",
      operatorLogonTime: new Date("2018-08-15T13:38:21.000Z"),
      bearing: 137.5,
      blockId: "G111-165",
      headwaySecs: 396.3,
      headwaySpacing: HeadwaySpacing.Ok,
      previousVehicleId: "y0620",
      scheduleAdherenceSecs: 0,
      scheduledHeadwaySecs: 420,
      isShuttle: false,
      isOverload: false,
      isOffCourse: false,
      isRevenue: true,
      layoverDepartureTime: 1576091880,
      dataDiscrepancies: [],
      stopStatus: {
        stopId: "stop",
        stopName: "Stop",
      },
      timepointStatus: {
        fractionUntilTimepoint: 0.0,
        timepointId: "TIMEP",
      },
      scheduledLocation: {
        routeId: "111",
        directionId: 1,
        tripId: "42199996",
        runId: "scheduled run",
        timeSinceTripStartTime: 0,
        headsign: "scheduled headsign",
        viaVariant: "5",
        timepointStatus: {
          timepointId: "TIMEP",
          fractionUntilTimepoint: 0.0,
        },
      },
      routeStatus: "pulling_out",
      endOfTripType: "another_trip",
      blockWaivers: [],
      crowding: null,
    }

    const tree = renderer
      .create(
        <IncomingBox
          vehiclesAndGhosts={[vehicle]}
          ladderDirection={LadderDirection.ZeroToOne}
          selectedVehicleId={undefined}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a ghost", () => {
    const ghost: Ghost = ghostFactory.build({
      id: "ghost",
      directionId: 0,
      routeId: "111",
      tripId: "42199995",
      headsign: "Woodlawn",
      blockId: "G111-165",
      runId: "126-1056",
      viaVariant: "5",
      layoverDepartureTime: 1576091880,
      scheduledTimepointStatus: {
        timepointId: "TIMEP",
        fractionUntilTimepoint: 0.0,
      },
      scheduledLogonTime: null,
      routeStatus: "pulling_out",
      blockWaivers: [],
    })

    const tree = renderer
      .create(
        <IncomingBox
          vehiclesAndGhosts={[ghost]}
          ladderDirection={LadderDirection.ZeroToOne}
          selectedVehicleId={undefined}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a crowding view of a vehicle", () => {
    const vehicle: Vehicle = {
      id: "y0654",
      label: "0654",
      runId: "126-1056",
      timestamp: 1576091706,
      latitude: 42.36296,
      longitude: -71.05814,
      directionId: 0,
      routeId: "111",
      tripId: "42199995",
      headsign: "Woodlawn",
      viaVariant: "5",
      operatorId: "1",
      operatorFirstName: "OPPIE",
      operatorLastName: "OPERATOR",
      operatorLogonTime: new Date("2018-08-15T13:38:21.000Z"),
      bearing: 137.5,
      blockId: "G111-165",
      headwaySecs: 396.3,
      headwaySpacing: HeadwaySpacing.Ok,
      previousVehicleId: "y0620",
      scheduleAdherenceSecs: 0,
      scheduledHeadwaySecs: 420,
      isShuttle: false,
      isOverload: false,
      isOffCourse: false,
      isRevenue: true,
      layoverDepartureTime: 1576091880,
      dataDiscrepancies: [],
      stopStatus: {
        stopId: "stop",
        stopName: "Stop",
      },
      timepointStatus: {
        fractionUntilTimepoint: 0.0,
        timepointId: "TIMEP",
      },
      scheduledLocation: {
        routeId: "111",
        directionId: 1,
        tripId: "42199996",
        runId: "scheduled run",
        timeSinceTripStartTime: 0,
        headsign: "scheduled headsign",
        viaVariant: "5",
        timepointStatus: {
          timepointId: "TIMEP",
          fractionUntilTimepoint: 0.0,
        },
      },
      routeStatus: "pulling_out",
      endOfTripType: "another_trip",
      blockWaivers: [],
      crowding: {
        occupancyStatus: "EMPTY",
        load: 0,
        occupancyPercentage: 0.0,
        capacity: 18,
      },
    }

    const tree = renderer
      .create(
        <IncomingBox
          vehiclesAndGhosts={[vehicle]}
          ladderDirection={LadderDirection.ZeroToOne}
          selectedVehicleId={undefined}
          displayCrowding={true}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("renders a crowding view missing crowding data", () => {
    const vehicle: Vehicle = {
      id: "y0654",
      label: "0654",
      runId: "126-1056",
      timestamp: 1576091706,
      latitude: 42.36296,
      longitude: -71.05814,
      directionId: 0,
      routeId: "111",
      tripId: "42199995",
      headsign: "Woodlawn",
      viaVariant: "5",
      operatorId: "1",
      operatorFirstName: "OPPIE",
      operatorLastName: "OPERATOR",
      operatorLogonTime: new Date("2018-08-15T13:38:21.000Z"),
      bearing: 137.5,
      blockId: "G111-165",
      headwaySecs: 396.3,
      headwaySpacing: HeadwaySpacing.Ok,
      previousVehicleId: "y0620",
      scheduleAdherenceSecs: 0,
      scheduledHeadwaySecs: 420,
      isShuttle: false,
      isOverload: false,
      isOffCourse: false,
      isRevenue: true,
      layoverDepartureTime: 1576091880,
      dataDiscrepancies: [],
      stopStatus: {
        stopId: "stop",
        stopName: "Stop",
      },
      timepointStatus: {
        fractionUntilTimepoint: 0.0,
        timepointId: "TIMEP",
      },
      scheduledLocation: {
        routeId: "111",
        directionId: 1,
        tripId: "42199996",
        runId: "scheduled run",
        timeSinceTripStartTime: 0,
        headsign: "scheduled headsign",
        viaVariant: "5",
        timepointStatus: {
          timepointId: "TIMEP",
          fractionUntilTimepoint: 0.0,
        },
      },
      routeStatus: "pulling_out",
      endOfTripType: "another_trip",
      blockWaivers: [],
      crowding: null,
    }

    const tree = renderer
      .create(
        <IncomingBox
          vehiclesAndGhosts={[vehicle]}
          ladderDirection={LadderDirection.ZeroToOne}
          selectedVehicleId={undefined}
          displayCrowding={true}
        />
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })

  test("clicking an incoming crowding icon selects the associated vehicle", () => {
    const mockDispatch = jest.fn()

    const vehicle: Vehicle = vehicleFactory.build()

    const wrapper = mount(
      <StateDispatchProvider state={initialState} dispatch={mockDispatch}>
        <IncomingBox
          vehiclesAndGhosts={[vehicle]}
          ladderDirection={LadderDirection.ZeroToOne}
          selectedVehicleId={undefined}
          displayCrowding={true}
        />
      </StateDispatchProvider>
    )
    wrapper.find(".m-incoming-box__vehicle").simulate("click")

    expect(mockDispatch).toHaveBeenCalledWith(selectVehicle(vehicle))
  })
})
