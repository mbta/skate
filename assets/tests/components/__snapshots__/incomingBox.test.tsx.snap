// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IncomingBox renders a ghost 1`] = `
<div
  className="m-incoming-box"
>
  <button
    className="m-incoming-box__vehicle "
    onClick={[Function]}
  >
    <svg
      style={
        Object {
          "height": 15.2,
          "width": 16.72,
        }
      }
      viewBox="-8.36 -7.6 16.72 15.2"
    >
      <g
        className="m-vehicle-icon m-vehicle-icon--small ghost"
      >
        <g
          transform="scale(0.26599999999999996) translate(-24,-23)"
        >
          <path
            className="m-vehicle-icon__ghost-highlight"
            d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
            stroke-join="round"
          />
          <path
            className="m-vehicle-icon__ghost-body"
            d="m43.79 19c0-9.68-8.79-17.49-19.59-17.49s-19.6 7.81-19.6 17.49v12.88 11a2 2 0 0 0 2.55 1.87l6.78-4.09 10.27 5.92 10.26-5.88 6.78 4.09a2 2 0 0 0 2.55-1.87z"
            stroke-join="round"
          />
        </g>
        <text
          className="m-vehicle-icon__variant"
          dominantBaseline="alphabetic"
          textAnchor="middle"
          x={0}
          y={3.5999999999999996}
        >
          5
        </text>
      </g>
    </svg>
    <div
      className="m-incoming-box__vehicle-label"
    >
      1056
    </div>
  </button>
</div>
`;

exports[`IncomingBox renders a vehicle 1`] = `
<div
  className="m-incoming-box"
>
  <button
    className="m-incoming-box__vehicle "
    onClick={[Function]}
  >
    <svg
      style={
        Object {
          "height": 15.2,
          "width": 16.72,
        }
      }
      viewBox="-8.36 -7.6 16.72 15.2"
    >
      <g
        className="m-vehicle-icon m-vehicle-icon--small on-time"
      >
        <path
          className="m-vehicle-icon__triangle"
          d="m27.34 9.46 16.84 24.54a4.06 4.06 0 0 1 -1 5.64 4.11 4.11 0 0 1 -2.3.71h-33.72a4.06 4.06 0 0 1 -4.06-4.11 4 4 0 0 1 .72-2.24l16.84-24.54a4.05 4.05 0 0 1 5.64-1.05 4 4 0 0 1 1.04 1.05z"
          transform="scale(0.38) rotate(0) translate(-24,-22)"
        />
        <text
          className="m-vehicle-icon__variant"
          dominantBaseline="alphabetic"
          textAnchor="middle"
          x={0}
          y={5.6}
        >
          5
        </text>
      </g>
    </svg>
    <div
      className="m-incoming-box__vehicle-label"
    >
      1056
    </div>
  </button>
</div>
`;

exports[`IncomingBox renders empty state 1`] = `
<div
  className="m-incoming-box"
/>
`;
